pkg/drift/terraform/terraform.go

```diff
@@ -0,0 +1,179 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terraform
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/abcxyz/guardian/pkg/drift/assets"
+	"github.com/abcxyz/guardian/pkg/drift/gcs"
+	"github.com/abcxyz/guardian/pkg/drift/iam"
+)
+
+type ResourceInstances struct {
+	Attributes struct {
+		ID      string   `json:"id"`
+		Members []string `json:"members,omitempty"`
+	}
+}
+
+type TerraformState struct {
+	Resources []struct {
+		Type      string              `json:"type"`
+		Instances []ResourceInstances `json:"instances"`
+	} `json:"resources"`
+}
+
+type Parser struct {
+	gcs             *gcs.Client
+	gcpAssetsByID   map[int64]assets.HierarchyNode
+	gcpAssetsByName map[string]assets.HierarchyNode
+	organizationID  int64
+}
+
+func NewParser(
+	ctx context.Context,
+	organizationID int64,
+	gcpFolders []assets.HierarchyNode,
+	gcpProjects []assets.HierarchyNode,
+) (*Parser, error) {
+	client, err := gcs.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing gcs Client: %w", err)
+	}
+	allAssets := append(gcpFolders, gcpProjects...)
+	assetsByID := make(map[int64]assets.HierarchyNode)
+	assetsByName := make(map[string]assets.HierarchyNode)
+	for _, a := range allAssets {
+		assetsByID[a.ID] = a
+	}
+	for _, a := range allAssets {
+		assetsByName[a.Name] = a
+	}
+
+	return &Parser{
+		gcs:             client,
+		gcpAssetsByID:   assetsByID,
+		gcpAssetsByName: assetsByName,
+		organizationID:  organizationID,
+	}, nil
+}
+
+func (p *Parser) GetAllTerraformStateFileURIs(ctx context.Context, gcsBuckets []string) ([]string, error) {
+	gcsURIs := []string{}
+	for _, bucket := range gcsBuckets {
+		allStateFiles, err := p.gcs.GetAllFilesWithName(ctx, bucket, "default.tfstate")
+		if err != nil {
+			return nil, fmt.Errorf("could not determine state files in GCS bucket %s: %w", bucket, err)
+		}
+		gcsURIs = append(gcsURIs, allStateFiles...)
+	}
+	return gcsURIs, nil
+}
+
+func (p *Parser) ProcessAllTerraformStates(ctx context.Context, gcsUris []string) ([]iam.AssetIAM, error) {
+	iams := []iam.AssetIAM{}
+	for _, uri := range gcsUris {
+		state := TerraformState{}
+		data, err := p.gcs.DownloadFileIntoMemory(ctx, uri)
+		if err != nil {
+			return nil, fmt.Errorf("failed to download gcs URI for terraform: %w", err)
+		}
+		if err = json.Unmarshal(data, &state); err != nil {
+			return nil, fmt.Errorf("failed to parse terraform state into json: %w", err)
+		}
```

Comment:
It's _really_ inefficient to download the entire object into a buffer, then unmarshal that buffer. It's better to stream the writer from the DownloadFileIntoMemory directly into a json.Decoder to save a lot of allocation and memory.